### DISCUSSION TITLE
Sidecar file for raw and existing jpeg

### DISCUSSION THREAD

**Discussion by b10102016** [photoprism/photoprism#1684] [Q&A] [answered]
Hi

May I know how the photoprism to handle existing jpeg for a raw?
Will photoprism try to generate sidecar jpeg for the raw without using the existing one?
If using the existing one, what is the rule of name?

ex:
XXX/xxx.raw
YYY/20211029-xxx.jpg

Will raw conversion still happen and generate the sidecar jpeg for xxx.raw?
Or the photoprism use the existing `20211029-xxx.jpg`?

Sincerely,
Ian

**Comment by lastzero** [comment]
If there's an existing JPEG, PhotoPrism won't create a new one.

**Comment by graciousgrey** [comment, marked as answer]
At the moment, the JPG and RAW file must be located in the same folder and share the same name, so that no additional JPG is created by PhotoPrism.

e.g. XXX/xxx.raw and XXX/xxx.jpg

Otherwise the indexer would need to scan the whole library for a matching JPG each time it encounters a RAW file.

**Reply by b10102016** [reply]
Thanks for your quick reply.

Really appreciate

**Reply by valpackett** [reply]
> Otherwise the indexer would need to scan the whole library for a matching JPG each time it encounters a RAW file.

No, it wouldn't need to *scan* many times, because hashmaps exist. Scan the filesystem once, build a map of basename → full_paths, here you have your stacks.

**Reply by lastzero** [reply]
So a two pass scan. At the moment, a one-pass scan is implemented. And how do you ultimately know that the files belong together? Do you think it's worth adding more complexity?

**Reply by valpackett** [reply]
> how do you ultimately know that the files belong together

First compare the basename, e.g. `Exported/IMG_20210203_123456.jpg` and `Raw/IMG_20210203_123456.ARW` have the same basename `IMG_20210101_123456`.

Now, unfortunately not everyone uses [a rename script](https://github.com/unrelentingtech/dotfiles/blob/master/bin/photo-rename) so they might end up with e.g. `2020/IMG_1234.jpg` and `2021/IMG_1234.CR2` as actually unrelated photos :/

But to resolve that, it is possible to analyze metadata — to confirm they are related, compare exif dates.

**Reply by lastzero** [reply]
We already compare metadata. The main issues are performance and matching files without the same file path, a proper unique ID, or the exact same time & location.
